middleware/complexconfig/provider.go

```diff
@@ -0,0 +1,178 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package complexconfig adds middleware for schema informed complex configuration
+// encoding/decoding as a work-around for https://github.com/pulumi/pulumi/pull/15032.
+//
+// The entry point for this package is [Wrap].
+//
+// Deprecated: This package will be removed after
+// https://github.com/pulumi/pulumi/pull/15032 merges.
+package complexconfig
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/pulumi/pulumi-go-provider/internal/putil"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/sig"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+
+	p "github.com/pulumi/pulumi-go-provider"
+)
+
+func Wrap(provider p.Provider) p.Provider {
+	encoder := provider
+	contract.Assertf(provider.GetSchema != nil, "provider.GetSchema must be implemented")
+	encoder.CheckConfig = encodeCheckConfig(provider.CheckConfig, provider.GetSchema)
+	return encoder
+}
+
+type (
+	getSchema   = func(context.Context, p.GetSchemaRequest) (p.GetSchemaResponse, error)
+	checkConfig = func(context.Context, p.CheckRequest) (p.CheckResponse, error)
+)
+
+func encodeCheckConfig(check checkConfig, getSchema getSchema) checkConfig {
+	if check == nil {
+		check = func(_ context.Context, req p.CheckRequest) (p.CheckResponse, error) {
+			return p.CheckResponse{Inputs: req.News}, nil
+		}
+	}
+	return func(ctx context.Context, req p.CheckRequest) (p.CheckResponse, error) {
+		// Only req.News is from the engine. req.Olds (if it exists) is from state
+		// and thus went through a previous normalizing pass of CheckConfig.
+
+		// If there are no inputs, then we can just return.
+		if len(req.News) == 0 {
+			return check(ctx, req)
+		}
+
+		schemaResp, err := getSchema(ctx, p.GetSchemaRequest{})
+		if err != nil {
+			return p.CheckResponse{},
+				p.InternalErrorf("unable to decode config: no schema available: %w", err)
+		}
+		var spec schema.PackageSpec
+		if err := json.Unmarshal([]byte(schemaResp.Schema), &spec); err != nil {
+			return p.CheckResponse{},
+				p.InternalErrorf("unable to decode config: invalid schema: %w", err)
+		}
+
+		for k, spec := range spec.Config.Variables {
+			v, ok := req.News[resource.PropertyKey(k)]
+			if !ok {
+				continue
+			}
+
+			req.News[resource.PropertyKey(k)] = fixEncoding(v, spec)
+		}
+
+		return check(ctx, req)
+	}
+}
+
+func fixEncoding(v resource.PropertyValue, spec schema.PropertySpec) (out resource.PropertyValue) {
+	// Ensure that v is unwrapped
+
+	if putil.IsComputed(v) {
+		return v
+	}
+
+	if putil.IsSecret(v) {
+		defer func() {
+			out = putil.MakeSecret(v)
+		}()
+		v = putil.MakePublic(v)
+	}
+
+	if !v.IsString() || spec.Type == "string" {
```

Comment:
This seems like a bug at first glance - if it's _not_ a string or if it _is_ a string. But I think I get it now: if the value is not a string, we can't unmarshal it. If the value is string-typed in the spec, there's no point unmarshaling it since it's a primitive string.

Reply:
Yes! I see how that could be confusing though. I'll add ☝️ as a comment.